### DISCUSSION TITLE
[help] Advice on highly heterogeneous workflows

### DISCUSSION THREAD

**Discussion by connor-duffin** [ropensci/targets#1400] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Our workflow is highly heterogeneous, ranging from single targets that use lots of memory on their own, to dynamically branched targets that are embarrassingly parallel. I have set up heterogeneous workers for the different parts of the pipeline on their own, but it seems they all run at the same time - in reality I was hoping that this would separate out the running of the pipeline, so that we don't run out of memory on our machine!

As a quickfix I am running something like:

``` r
targets::tar_make("highly_parallel_target")  # with lots of branches
targets::tar_make("expensive_target")
targets::tar_make()  # remainder of the targets
```

I was wondering if there was any advice on dealing with such a pipeline, and, if any, best practices that should be followed.

**Comment by wlandau** [comment, marked as answer]
Successive `tar_make()` calls are totally fine. If you prefer a single one, then you might consider restructuring the dependency graph to force certain targets to run after one another. For example, instead of:

```r
library(targets)
list(
  tar_target(name = expensive, command = f()),
  tar_target(name = cheap, command = g())
)
```

![Screenshot 2024-12-17 at 9 39 36 AM](https://github.com/user-attachments/assets/3d84c75f-698a-4147-9e72-6a5ca0a21bdf)

you could consider something like:

```r
library(targets)
list(
  tar_target(name = expensive, command = f()),
  tar_target(
    name = cheap,
    command = {
      expensive
      g()
    }
  )
)
```

![Screenshot 2024-12-17 at 9 40 43 AM](https://github.com/user-attachments/assets/cc6e9549-9397-49bb-9349-c67b2ec1985c)

or a different variation of this that doesn't load `expensive` inside `cheap`:

```r
library(targets)
list(
  tar_target(name = expensive, command = f()),
  tar_target(
    name = sentinel,
    command = {
      expensive
      NULL
    }
  ),
  tar_target(
    name = cheap,
    command = {
      sentinel
      g()
    }
  )
)
```

**Reply by connor-duffin** [reply]
Great - that all makes sense to me and is very helpful.
